cmd/restic/cmd_rebuild_index.go

```diff
@@ -44,59 +52,100 @@ func runRebuildIndex(gopts GlobalOptions) error {
 		return err
 	}
 
-	ctx, cancel := context.WithCancel(gopts.ctx)
-	defer cancel()
-	return rebuildIndex(ctx, repo, restic.NewIDSet())
+	return rebuildIndex(opts, gopts, repo, restic.NewIDSet())
 }
 
-func rebuildIndex(ctx context.Context, repo restic.Repository, ignorePacks restic.IDSet) error {
-	Verbosef("counting files in repo\n")
+func rebuildIndex(opts RebuildIndexOptions, gopts GlobalOptions, repo *repository.Repository, ignorePacks restic.IDSet) error {
+	ctx := gopts.ctx
+
+	var obsolete restic.IDs
+	packSizeFromList := make(map[restic.ID]int64)
+	packs := restic.NewIDSet()
+	totalPacks := 0
+
+	if opts.ReadAllPacks {
+		// get old index files
+		err := repo.List(ctx, restic.IndexFile, func(id restic.ID, size int64) error {
+			obsolete = append(obsolete, id)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
 
-	var packs uint64
-	err := repo.List(ctx, restic.PackFile, func(restic.ID, int64) error {
-		packs++
-		return nil
-	})
-	if err != nil {
-		return err
-	}
+		Verbosef("finding pack files in repo...\n")
+		err = repo.List(ctx, restic.PackFile, func(id restic.ID, size int64) error {
+			packSizeFromList[id] = size
+			packs.Insert(id)
+			totalPacks++
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	} else {
+		Verbosef("loading indexes...\n")
+		err := repo.LoadIndex(gopts.ctx)
+		if err != nil {
+			return err
+		}
 
-	bar := newProgressMax(!globalOptions.Quiet, packs-uint64(len(ignorePacks)), "packs")
-	idx, invalidFiles, err := index.New(ctx, repo, ignorePacks, bar)
-	bar.Done()
-	if err != nil {
-		return err
-	}
+		packSizeFromIndex := make(map[restic.ID]int64)
```

Comment:
Please extract the `packSizeFromIndex` calculation into a separate method. #3082 also adds the exact same loop.

Reply:
Fully agree. In fact I would add this to `repository.MasterIndex`.

There is one thing we should consider about this pack size computation used here, in #3082 and in the new `prune` implementations: In the discussions about a repo format V2 there is the proposal to save the pack header in JSON format and I would strongly advice to do so (e.g. to support compression :wink: ). However, this would make the computation of the pack header size much more difficult and even impossible if old and new pack format formats will be supported simultaneously. So IMHO saving the pack header in JSON would also imply that we overwork the format of the index files and save some additional pack information there - like the pack header size or whether the pack header is in JSON or not. Therefore this information anyway is much better suited to be saved within the in-memory index (and if we enlarge the pack files, there won't be much of an memory issue).

Do you mind if I keep this comparably small piece of code as it is in this PR and prepare a follow-up PR which extracts this calculation and prepares for a maybe-coming additional pack file format?

Reply:
To handle compressed blobs the index must be changed. We'd need both the compressed and uncompressed blob sizes in the index. And along with that we can simply add enough information to allow check and the other commands to deduce the pack header size. I'm not too much a fan of JSON in particular as parsing it is slow as hell. (And without additional header compression that JSON would be two or three times larger than the current binary format, for the same information!)

Deduplicating that pack size loop in a separate PR is fine for me.

Reply:
>  We'd need both the compressed and uncompressed blob sizes in the index.

@MichaelEischer Just as a side note: I think we need to save the (uncompressed, unencrypted) blob size and the size the blob occupies within the pack (i.e. maybe compressed and including crypto overhead) in the index.